doc: revise zone concierge doc (round 2) (#1234)
babylonlabs-io/babylon: x/zoneconcierge/README.md

# ZoneConcierge

The Zone Concierge module is responsible for providing BTC staking integration
functionalities for Bitcoin Supercharged Networks (BSNs) connecting  
to Babylon Genesis through IBC. It leverages the IBC protocol to receive BSNs'
headers, and propagate BTC timestamps of those headers and information
associated with the BTC staking protocol (e.g., finality providers, BTC stakes,
and more).  
The Zone Concierge module synchronises the following information with BSNs (aka
consumers) via IBC packets:

- **BTC Headers:** Babylon Genesis forwards the BTC headers maintained by its
  BTC light client to BSNs.  
  This allows BSNs to maintain an image of the Bitcoin chain and verify
  information included in it through inclusion proofs (e.g., an inclusion
  proof of a BTC Timestamp containing BSN headers).
- **BTC Timestamps:** When a Babylon Genesis epoch is finalized, the Babylon
  Genesis chain sends BTC timestamps to BSNs. Each BTC timestamp contains:
  - The latest BSN header that was checkpointed in the finalised epoch
  - Recent BTC headers that extend the BSN's BTC light client
  - The finalised epoch's metadata and raw checkpoint

>   - Proofs that the BSN header was included in the epoch and the epoch was
>     properly sealed and submitted to Bitcoin

Reviewer: ```suggestion
  - Proofs that a BSN header was included in an epoch and the epoch was
  timestamped on the Bitcoin chain.
```

message ConsumerSlashingIBCPacket {
  /// evidence is the FP slashing evidence that the Consumer sends to Babylon
  babylon.finality.v1.Evidence evidence = 1;
}
```

### Processing Inbound IBC Packets

The `HandleConsumerSlashing` function (called upon
[OnRecvPacket](x/zoneconcierge/module_ibc.go)) processes slashing reports
received from BSNs through IBC packets, with the following workflow:

1. **Verifying Evidence**: 
   - Validates that slashing evidence is present and well-formed
   - Extracts the BTC secret key from the evidence
   - Verifies that the finality provider's BTC public key matches the evidence
2. **Slashing Execution**:
   - Updates the BSN finality provider's slashed status
   - Sends power distribution update events to adjust the Babylon finality
     provider's voting power
   - Identifies all BTC delegations associated with the slashed finality
     provider
   - Creates slashed BTC delegation events for each affected BSN
   - Propagates the slashing event to each BSN
3. **Event Emission**: Emits a `EventSlashedFinalityProvider` event for external
   slashing mechanisms (e.g., BTC slasher/vigilante)

> 

Author: > how is timely propagation ensured 

The propagation is handled by IBC so the timing depends on the IBC protocol, e.g., how timely the relayer is picking up the IBC packets. It is not in this spec's scope imo

> and what's the algorithm to identify which BSNs need to receive an update and what are its performance considerations

Added more details in L451-452. The algorithm is roughly as follows: given a slashed FP, 

- For each BTC delegation
  - find all multi-staked FPs and their BSNs
  - For each of those BSNs, append the slashed BTC delegation event to a buffer of this BSN
- Send IBC packet with all those events to each of the BSN with non-empty buffer
